### PR TITLE
[map/scatter] create shared source selector component

### PR DESCRIPTION
**Pull request by chejennifer**
- create a source selector component to be used by map, scatter, and data download tool (will integrate into data download tool in later PR)
- also fixes [issue #1465](https://github.com/datacommonsorg/website/issues/1465)

for map tool:
screen-recording-_3_

for scatter tool:
screen-recording-_4_

### CODE REVIEW

File: static/js/shared/source_selector.tsx [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+const UP_ARROW_HTML = <i className="material-icons">expand_less</i>;
+const MINUS_HTML = <i className="material-icons">remove</i>;
+const PLUS_HTML = <i className="material-icons">add</i>;
+const SELECTOR_PREFIX = "source-selector";
+
+export interface SourceSelectorSvInfo {
+  dcid: string;
+  name: string;
+  metahash: string;
+  metadataMap: Record<string, StatMetadata>;
```
Comment (shifucun) [inline static/js/shared/source_selector.tsx:48]: comment on "metahash" and metadataMap

File: static/js/shared/source_selector.tsx [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+      const importName = svInfo.metadataMap[metahash].importName;
+      if (!importName) {
+        sourceOptionsNoImportName.push(metahash);
+        return;
+      }
+      if (!(importName in importNameToSourceOptions)) {
+        importNameToSourceOptions[importName] = [];
+      }
+      importNameToSourceOptions[importName].push(metahash);
+      if (importNameToSourceOptions[importName].length > 3) {
```
Comment (shifucun) [inline static/js/shared/source_selector.tsx:182]: make 3 a constant

File: static/js/shared/source_selector.tsx [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+  svInfo: SourceSelectorSvInfo[]
+): Record<string, string> {
+  const result = {};
+  svInfo.forEach((info) => (result[info.dcid] = info.metahash));
+  return result;
+}
+
+function getSourceTitle(metadata: StatMetadata): string {
+  if (_.isEmpty(metadata)) {
+    return "Best Available";
```
Comment (shifucun) [inline static/js/shared/source_selector.tsx:275]: We should add note somewhere on what is "Best Available"..

File: static/js/tools/scatter/util.ts [hunk 4]
```diff
@@ -73,6 +73,41 @@ async function getStatsWithinPlace(
   });
 }
 
+export async function getAllStatsWithinPlace(
```
Comment (beets) [inline static/js/tools/scatter/util.ts:76]: please add documentation, especially for shared functions

**Review by beets — COMMENTED** [review]
this is nice, the modal works well for this.

File: static/js/shared/source_selector.tsx (hunk 1 above)
Comment (chejennifer) [reply static/js/shared/source_selector.tsx:58]: done

File: static/js/shared/source_selector.tsx (hunk 2 above)
Comment (chejennifer) [reply static/js/shared/source_selector.tsx]: done

File: static/js/shared/source_selector.tsx (hunk 3 above)
Comment (chejennifer) [reply static/js/shared/source_selector.tsx]: done

File: static/js/shared/source_selector.tsx [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+  Label,
+  Modal,
+  ModalBody,
+  ModalFooter,
+  ModalHeader,
+} from "reactstrap";
+
+import { StatMetadata } from "../tools/shared_util";
+
+const MODAL_MAX_WIDTH = "700px";
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/shared/source_selector.tsx]: done

File: static/js/shared/source_selector.tsx [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+const DOWN_ARROW_HTML = <i className="material-icons">expand_more</i>;
+const UP_ARROW_HTML = <i className="material-icons">expand_less</i>;
+const MINUS_HTML = <i className="material-icons">remove</i>;
+const PLUS_HTML = <i className="material-icons">add</i>;
+const SELECTOR_PREFIX = "source-selector";
+
+export interface SourceSelectorSvInfo {
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/shared/source_selector.tsx:49]: done

File: static/js/shared/source_selector.tsx [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        <ModalFooter>
+          <Button color="primary" onClick={onConfirm}>
+            Update
+          </Button>
+        </ModalFooter>
+      </Modal>
+    </>
+  );
+
+  function getSourceOptionJsx(
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/shared/source_selector.tsx]: done

File: static/js/shared/source_selector.tsx [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+  opened: boolean,
+  svName: string,
+  selectedMetadata: StatMetadata
+): JSX.Element {
+  const sourceTitle = getSourceTitle(selectedMetadata);
+  return (
+    <div
+      className={
+        opened
+          ? `${SELECTOR_PREFIX}-trigger ${SELECTOR_PREFIX}-sv-trigger-opened`
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/shared/source_selector.tsx]: done

File: static/js/shared/source_selector.tsx [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+              <div>scalingFactor: {metadata.scalingFactor}</div>
+            )}
+            {metadata.unit && <div>unit: {metadata.unit}</div>}
+          </div>
+        </Label>
+      </FormGroup>
+    );
+  }
+
+  function getSourceOptionSectionJsx(
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/shared/source_selector.tsx]: done

File: static/js/tools/scatter/chart_loader.tsx [hunk 10]
```diff
@@ -179,15 +217,32 @@ async function loadData(
         `&stat_vars=${DEFAULT_POPULATION_DCID}`
     )
     .then((resp) => resp.data);
-  Promise.all([statResponsePromise, populationPromise])
-    .then(([statResponse, populationData]) => {
+  Promise.all([statResponsePromise, statAllResponsePromise, populationPromise])
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/tools/scatter/chart_loader.tsx:220]: I did not

File: static/js/tools/scatter/util.ts (hunk 4 above)
Comment (chejennifer) [reply static/js/tools/scatter/util.ts:83]: done

**Review by beets — APPROVED** [review]
thanks for the updates

**Pull request merged**